crates/emitter/src/control_structures.rs

```diff
@@ -0,0 +1,220 @@
+        self.emit_jump(emitter);
+        let depth = emitter.emit.stack_depth();
+
+        // The JITs rely on a jump target being emitted after the
+        // conditional jump
+        if self.should_fallthrough() {
+            emitter.emit.jump_target();
+        }
+        JumpPatchEmitter { offsets, depth }
+    }
+}
+
+// Compared to C++ impl, this uses explicit stack struct,
+// given Rust cannot store a reference of stack-allocated object into
+// another object that has longer-lifetime.
+//
+// Some methods are implemented here, instead of each EmitterScope.
+pub struct LoopStack {
+    loop_stack: Vec<LoopControl>,
+}
+
+impl LoopStack {
+    pub fn new() -> Self {
+        Self {
+            loop_stack: Vec::new(),
+        }
+    }
+
+    pub fn open_loop(&mut self, emit: &mut InstructionWriter) {
+        let depth = (self.loop_stack.len() + 1) as u8;
+        let new_loop = LoopControl::new(emit, depth);
+        self.loop_stack.push(new_loop);
+    }
+
+    pub fn register_break(&mut self, offset: BytecodeOffset) {
+        let innermost = self.innermost();
+        innermost.register_break(offset);
+    }
+
+    pub fn close_loop(&mut self, emit: &mut InstructionWriter) {
+        let innermost = self
+            .loop_stack
+            .pop()
+            .expect("There should be at least one loop");
+        innermost.emit_end_target(emit);
+    }
+
+    pub fn innermost(&mut self) -> &mut LoopControl {
+        self.loop_stack
+            .last_mut()
+            .expect("There should be at least one loop")
+    }
+}
+
+pub struct LoopControl {
+    pub breaks: Vec<BytecodeOffset>,
+    pub continues: Vec<BytecodeOffset>,
+    head: BytecodeOffset,
+}
+
+impl LoopControl {
+    pub fn new(emit: &mut InstructionWriter, depth: u8) -> Self {
+        let mut offset = emit.bytecode_offset();
+        if offset.offset == 0 {
+            emit.nop();
+            offset = emit.bytecode_offset();
+        }
+        emit.loop_head(depth);
+        // emit the jump target for the loop head
+        Self {
+            breaks: Vec::new(),
+            continues: Vec::new(),
+            head: offset,
+        }
+    }
+
+    pub fn register_break(&mut self, offset: BytecodeOffset) {
+        self.breaks.push(offset);
+    }
+
+    // TODO: fix continues so that they work with scopes correctly
+    // pub fn emit_continue_target(&mut self, emit: &mut InstructionWriter) {
+    //     if self.continues.len() > 0 {
+    //         emit.patch_and_emit_jump_target((*self.continues).to_vec());
+    //     }
+    // }
+
+    pub fn emit_end_target(self, emit: &mut InstructionWriter) {
+        let offset = emit.bytecode_offset();
+        emit.goto_(self.head.offset as i32);
+        emit.patch_jump_to_target(self.head, offset);
+
+        emit.patch_and_emit_jump_target(self.breaks);
+    }
+}
+
+// Struct for multiple jumps that point to the same target. Examples are breaks and continues.
```

Comment:
This is a bit vague what's this is for.

While this is named `break`, this is used also for loop condition.
Can you add comment about the loop condition use case?

also, I feel continue should be different class.